python/ray/data/_internal/logical/rules/randomize_blocks.py

```diff
@@ -71,6 +71,13 @@ def _apply(self, op: LogicalOperator) -> LogicalOperator:
 
         # Add RandomizeBlocks operator as the last operator in the DAG if necessary.
         for random_op in operators:
+            # Only add the RandomizeBlocks operator if it is not a duplicate.
+            if (
+                isinstance(op, RandomizeBlocks)
+                and op._seed is None
+                and random_op._seed is None
+            ):
+                continue
```

Comment:
Why is operators a list at all?

Reply:
good question. this is the existing implementation, I did the least intrusive modification here - but if you want to refactor it please feel free.

Reply:
given limited time constraints, best to keep refactoring out of scope (separate PR) unless if you say otherwise